File: leetcode/searching/153-find-minimum-in-rotated-sorted-array/c/solution.c
#include <stdlib.h>

/// {{
/// Problem: 153. Find Minimum in Rotated Sorted Array
/// Difficulty: `Medium`
/// Links: https://leetcode.com/problems/find-minimum-in-rotated-sorted-array/
/// Topics: `array`, `binary-search`
/// Timestamp: `Fri, 04 Jul 2025 18:17:54 +0530`

int findMin(int* nums, int numsSize) {
  int low = 0, high = numsSize - 1;
  int min = nums[low];

  while (low <= high) {
    int mid = (low + high) / 2;

    if (nums[low] <= nums[mid]) {
      if (nums[mid] <= nums[high]) {
        min = nums[low];
        break;
      } else {
        low = mid + 1;
      }
    } else {
      low = low + 1;
    }
  }

  return min;
}

/// }}

File: leetcode/searching/153-find-minimum-in-rotated-sorted-array/js/solution.js
/// {{
/// Problem: 153. Find Minimum in Rotated Sorted Array
/// Difficulty: `Medium`
/// Links: https://leetcode.com/problems/find-minimum-in-rotated-sorted-array/
/// Topics: `array`, `binary-search`
/// Timestamp: `Wed, 30 Jul 2025 01:57:13 +0530`

/**
 * @param {number[]} nums
 * @return {number}
 */
function findMin(nums) {
  const n = nums.length;
  let low = 0, high = n - 1;

  while (low < high) {
    const mid = Math.floor((low + high) / 2);

    if (nums[mid] <= nums[high]) {
      high = mid;
    } else {
      low = mid + 1;
    }
  }

  return nums[low];
};

// }}